File: python/exer02b02_producer_consumer.py
'''
THE PRODUCER-CONSUMER PROBLEM

SOLUTION TYPE B: USING SEMAPHORES
    Version B02: 2 slow producers, 1 fast consumer
'''

import time
import threading



def producer(
    sem_fill: threading.Semaphore,
    sem_empty: threading.Semaphore,
    q: list,
    start_value: int
):
    i = 1

    while True:
        sem_empty.acquire()
        q.append(i + start_value)
        time.sleep(1)
        sem_fill.release()
        i += 1



def consumer(
    sem_fill: threading.Semaphore,
    sem_empty: threading.Semaphore,
    q: list
):
    while True:
        sem_fill.acquire()
        data = q.pop(0)
        print('Consumer', data)
        sem_empty.release()



s_fill = threading.Semaphore(0)  # item produced
s_empty = threading.Semaphore(1) # remaining space in queue
que = []

threading.Thread(target=producer, args=(s_fill, s_empty, que, 0)).start()
threading.Thread(target=producer, args=(s_fill, s_empty, que, 1000)).start()

threading.Thread(target=consumer, args=(s_fill, s_empty, que)).start()

File: csharp/exercise/exer02b02-producer-consumer.cs
/*
 * THE PRODUCER-CONSUMER PROBLEM
 *
 * SOLUTION TYPE B: USING SEMAPHORES
 *      Version B02: 2 slow producers, 1 fast consumer
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Exer02B02 : IRunnable
{
    public void run()
    {
        var semFill = new Semaphore(0, int.MaxValue);     // item produced
        var semEmpty = new Semaphore(1, int.MaxValue);    // remaining space in queue

        var queue = new Queue<int>();

        new Thread(() => producer(semFill, semEmpty, queue, 0)).Start();
        new Thread(() => producer(semFill, semEmpty, queue, 1000)).Start();

        new Thread(() => consumer(semFill, semEmpty, queue)).Start();
    }


    private void producer(Semaphore semFill, Semaphore semEmpty,
        Queue<int> queue, int startValue)
    {
        int i = 1;

        for (; ; ++i)
        {
            semEmpty.WaitOne();

            queue.Enqueue(i + startValue);
            Thread.Sleep(1000);

            semFill.Release();
        }
    }


    private void consumer(Semaphore semFill, Semaphore semEmpty, Queue<int> queue)
    {
        for (; ; )
        {
            semFill.WaitOne();

            int data = queue.Dequeue();
            Console.WriteLine("Consumer " + data);

            semEmpty.Release();
        }
    }
}
